### PR TITLE
[DO NOT MERGE] try generic position

### PR DESCRIPTION
**Pull request by vincentsarago**
This is a try at solving (at least partially) https://github.com/developmentseed/geojson-pydantic/issues/98

```python
from geojson_pydantic.types import Position2D, Position3D
from geojson_pydantic import Point

p2 = Point[Position2D]
>> p2(type="Point", coordinates=(0,0,0))
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 p2(type="Point", coordinates=(0,0,0))

ValidationError: 1 validation error for Point[Tuple[float, float]]
coordinates
  wrong tuple length 3, expected 2 (type=value_error.tuple.length; actual_length=3; expected_length=2)

p2(type="Point", coordinates=(0,0))
>> Point[Tuple[float, float]](type='Point', coordinates=(0.0, 0.0))


p = Polygon[Position3D]
p(type="Polygon", coordinates=[[(1.0, 2.0), (3.0, 4.0), (5.0, 6.0), (1.0, 2.0)]])
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
Input In [21], in <cell line: 1>()
----> 1 p(type="Polygon", coordinates=[[(1.0, 2.0), (3.0, 4.0), (5.0, 6.0), (1.0, 2.0)]])

ValidationError: 4 validation errors for Polygon[Tuple[float, float, float]]
coordinates -> 0 -> 0
  wrong tuple length 2, expected 3 (type=value_error.tuple.length; actual_length=2; expected_length=3)
coordinates -> 0 -> 1
  wrong tuple length 2, expected 3 (type=value_error.tuple.length; actual_length=2; expected_length=3)
coordinates -> 0 -> 2
  wrong tuple length 2, expected 3 (type=value_error.tuple.length; actual_length=2; expected_length=3)
coordinates -> 0 -> 3
  wrong tuple length 2, expected 3 (type=value_error.tuple.length; actual_length=2; expected_length=3)
```

@eseglem what do you think about this? 

cc @geospatial-jeff

### CODE REVIEW

File: geojson_pydantic/types.py [hunk 1]
```diff
@@ -1,24 +1,11 @@
...
-    LineStringCoords = conlist(Position, min_items=2)
-    LinearRing = conlist(Position, min_items=4)
-
-MultiPointCoords = List[Position]
-MultiLineStringCoords = List[LineStringCoords]
-PolygonCoords = List[LinearRing]
-MultiPolygonCoords = List[PolygonCoords]
+Position2D = Tuple[float, float]
+Position3D = Tuple[float, float, float]
+Position = Union[Position2D, Position3D]
```
Comment (vincentsarago) [inline geojson_pydantic/types.py:11]: we could get rid of the `types.py` files if we go this way

**Comment by eseglem** [comment]
I haven't done much with generics myself. Does this end up with something like `Feature[Point[Position2D], MyProperties]`? And everything still works?

Would we also provide Goemetry2D and Feature2D or is that going to be left to an end User? 

I'm not necessarily a fan of preventing the Z, based on my reading of the spec. I totally understand the usefulness and desire to have it though. Heck I might even use it after debating myself about it for way too long.

I was going to caution against moving the types because it's a breaking change if anyone uses them. But the next release will be have breaking changes anyways, so seems like a reasonable time to do it.

And while I'm thinking of it and it's semi relivant: Why doesn't GeometryCollection inherit from a shared base with Geometry? They share most of the same functions and there's a little unnecessary duplication I think.

**Comment by vincentsarago** [comment]
> Would we also provide Goemetry2D and Feature2D or is that going to be left to an end User?

yep this was going to be my next commit

**Comment by eseglem** [comment]
I believe there are some weird limitation with `Annotated` being nested
inside stuff.  I ran into some of that when trying to discriminate between
Geometry and GeometryCollection. I can't remember the exact details. But
one of those things that pretty much can't / won't be fixed until
pydantic2.

I kind of expected things would end up real weird / nasty nesting generic
inside generic with discriminator and such. It's right around the limits of
pydantic. I'll try to play with a few things if I have time tonight though,
see if I can come up with something.

**Comment by vincentsarago** [comment]
maybe instead of annotated type we could go back to a simpler method https://github.com/developmentseed/geojson-pydantic/issues/93#issuecomment-1404914825 🙈

**Comment by vincentsarago** [comment]
I removed the annotated type in last commit (so reverting to verbose/useless  validation error message) just to see if it would work with python 3.8 ... but it doesn't 

I've tried other things locally without much luck. I'm going to pause this work and maybe re-check when pydantic v2 will be out

**Pull request closed without merging**